src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveRenderedService.cs

```diff
@@ -0,0 +1,127 @@
+public class RetrieveRenderedService : IRetrieveRenderedService
+{
+    private readonly IFileStore _blobDataStore;
+    private readonly IInstanceStore _instanceStore;
+    private readonly IDicomRequestContextAccessor _dicomRequestContextAccessor;
+    private readonly RetrieveConfiguration _retrieveConfiguration;
+    private readonly RecyclableMemoryStreamManager _recyclableMemoryStreamManager;
+    private readonly ILogger<RetrieveRenderedService> _logger;
+
+    public RetrieveRenderedService(
+        IInstanceStore instanceStore,
+        IFileStore blobDataStore,
+        IDicomRequestContextAccessor dicomRequestContextAccessor,
+        IOptionsSnapshot<RetrieveConfiguration> retrieveConfiguration,
+        RecyclableMemoryStreamManager recyclableMemoryStreamManager,
+        ILogger<RetrieveRenderedService> logger)
+    {
+        EnsureArg.IsNotNull(instanceStore, nameof(instanceStore));
+        EnsureArg.IsNotNull(blobDataStore, nameof(blobDataStore));
+        EnsureArg.IsNotNull(dicomRequestContextAccessor, nameof(dicomRequestContextAccessor));
+        EnsureArg.IsNotNull(retrieveConfiguration?.Value, nameof(retrieveConfiguration));
+        EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        _instanceStore = instanceStore;
+        _blobDataStore = blobDataStore;
+        _dicomRequestContextAccessor = dicomRequestContextAccessor;
+        _retrieveConfiguration = retrieveConfiguration?.Value;
+        _recyclableMemoryStreamManager = recyclableMemoryStreamManager;
+        _logger = logger;
+    }
+
+    public async Task<RetrieveRenderedResponse> RetrieveRenderedImageAsync(RetrieveRenderedRequest request, CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+        var partitionKey = _dicomRequestContextAccessor.RequestContext.GetPartitionKey();
+
+        try
+        {
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            AcceptHeader returnHeader = GetValidRenderAcceptHeader(request.AcceptHeaders);
+
+            // this call throws NotFound when zero instance found
+            InstanceMetadata instance = (await _instanceStore.GetInstancesWithProperties(
+                ResourceType.Instance, partitionKey, request.StudyInstanceUid, request.SeriesInstanceUid, request.SopInstanceUid, cancellationToken)).First();
+
+            FileProperties fileProperties = await RetrieveHelpers.CheckFileSize(_blobDataStore, _retrieveConfiguration.MaxDicomFileSize, instance, cancellationToken);
+            Stream stream = await _blobDataStore.GetFileAsync(instance.VersionedInstanceIdentifier, cancellationToken);
+            DicomFile dicomFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand);
+            DicomPixelData dicomPixelData = dicomFile.GetPixelDataAndValidateFrames(new[] { request.FrameNumber });
```

Comment:
Why this is needed now? The other comment says we rely on IndexOutOfRangeException

Reply:
Updated